README.md
# Blauberg Ventilation Integration for Home Assistant
[![MIT license][license-badge]][license-url]
[![Python][python-badge]][python-url]
[![Docker][docker-badge]][docker-url]

This repository provides an integration service for the Blauberg ventilation system with Home Assistant, specifically designed for the Blauberg Modbus S21 controller. It scrapes the counters from the S21 controller via Modbus, and publishes metrics to MQTT on separate topics, while also publishing configuration profiles for Home Assistant. The service is developed in Python and is suitable for use with Docker.

## Blauberg Series Supported

The following Blauberg Ventilation series are compatible with this service, provided they use the Modbus S21 controller:

- Serie KOMFORT


## Features

- Scrapes counters from the S21 controller via Modbus.
- Publishes metrics to MQTT separate topics, each of them having a predefined description, icon, and measure units.
- Publishes configuration profiles for Home Assistant.
- Enables control of the Blauberg ventilation system through switches and selects in Home Assistant.

The service also provides an API with read-only access to the following data types from the Blauberg ventilation system:

- Alarms
- Coils
- Discrete Inputs
- Input Registers
- Holding Registers

Note: The API does not provide methods to control these data types. It is designed only for retrieving their current state.

## Prerequisites

- Docker

## Installation

Create a directory for the project and create a `config.ini` file inside it:

```bash
mkdir blauberg-homeassistant
cd blauberg-homeassistant
touch config.ini
```

Create a `docker-compose.yml` file inside the directory:

```bash
touch docker-compose.yml
```

Copy the following content into the `docker-compose.yml` file:

```yaml
version: "2"
services:
  blauberg-homeassistant:
    image: gracingpro/blauberg-homeassistant:latest
    container_name: blauberg-homeassistant
    network_mode: bridge
    extra_hosts:
      - "host.docker.internal:host-gateway"
    ports:
      - "8006:8006"
    volumes:
      - ./config.ini:/opt/blauberg/config.ini
    restart: always
```

Copy the following content into the `config.ini` file:

```ini
[Vent]
ip = 192.168.0.251

[MQTT]
user = mqtt_user
topic = Blauberg
homeassistant_topic = homeassistant
password = mqtt_password
host = host.docker.internal
port = 1883
scrape_interval = 15

[Blauberg]
api_token = 123
api_port = 8006
```

Here is a description of each section and the individual properties:

- `[Vent]`
  - `ip`: This is the IP address of the Blauberg ventilation unit.

- `[MQTT]`
  - `user`: The username for authentication with the MQTT broker.
  - `topic`: The main topic under which the service will publish messages.
  - `homeassistant_topic`: The topic that the Home Assistant listens to for the configuration of the Blauberg component.
  - `password`: The password for authentication with the MQTT broker.
  - `host`: The IP address of the MQTT broker.
  - `port`: The port on which the MQTT broker is running.
  - `scrape_interval`: The interval, in seconds, at which the service will scrape data from the Blauberg ventilation unit and publish it to the MQTT broker.

- `[Blauberg]`
  - `api_token`: The API token to authenticate with the Blauberg API.
  - `api_port`: The port on which the Blauberg API is running.

This configuration file is essential for the operation of the service. Without it, the application wouldn't know how to connect to the ventilation unit, the MQTT broker, or the Blauberg API. Always ensure that this file is correctly configured before running the application.

## Usage

To start the service, run the following command:

```bash
docker-compose up -d
```

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

## Versioning and Changelog
Blauberg Ventilation Integration for Home Assistant follows semantic versioning. Check out the [Change Log](CHANGELOG.md) for information on the latest updates and improvements.


## Licensing

Blauberg Ventilation Integration for Home Assistant is licensed under [MIT License](https://opensource.org/license/mit/). For more details, see the [License file](LICENSE.md).



[license-url]: LICENSE.md
[license-badge]: https://img.shields.io/github/license/exness/mock-xhr-request.svg
[python-badge]: https://img.shields.io/badge/python-3.11%2B-blue.svg
[python-url]: https://www.python.org/downloads/release/python-3110/
[docker-badge]: https://img.shields.io/badge/docker-20.10.8-blue.svg
[docker-url]: https://www.docker.com/


CHANGELOG.md
# Changelog

All notable changes to the "Blauberg Ventilation Integration for Home Assistant" project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.0.0] - 2023-05-28

### Added
- Initial commit of the service, which includes basic integration features like Modbus connection, MQTT publishing and Home Assistant configuration profile publication.


## [1.0.1] - 2023-06-02

### Fixed
- Fix: Correct parsing for hour metrics in Modbus registers.


## [1.0.2] - 2023-06-03

### Added
- Add: Enhanced logging functionality

## [1.0.3] - 2023-06-16

### Fixed
- Fix: Fixed Monitor Stuck.

## [1.0.4] - 2023-07-07
- Fix: Fixed IR_VerMAIN_FMW incorrect icon.

## [1.0.5] - 2023-12-24
- Add: Support for two registers related to bypass functionality

Data/MODBUS.py
from pyModbusTCP.client import ModbusClient
from References.Descriptions.InputRegisters import input_registers as description_input_registers
from References.Descriptions.HoldingRegisters import holding_registers as description_holding_registers
from References.Descriptions.Coils import coils as description_coils
from References.Descriptions.DiscreteInputs import discrete_inputs as description_discrete_inputs
from References.Descriptions.Alarms import alarms as description_alarms
from References.Variables.InputRegisters import input_registers as variables_input_registers
from References.Variables.HoldingRegisters import holding_registers as variables_holding_registers
from References.Variables.Coils import coils as variables_coils
from References.Variables.DiscreteInputs import discrete_inputs as variables_discrete_inputs
from References.Variables.Alarms import alarms as variables_alarms
from References.Units.InputRegisters import input_registers as units_input_registers
from References.Units.HoldingRegisters import holding_registers as units_holding_registers
from References.Controls.Coils import coils as control_coils
from References.Controls.HoldingRegisters import holding_registers as control_holding_registers
from References.Aliases.HoldingRegisters import holding_registers as aliases_holding_registers
from References.Aliases.InputRegisters import input_registers as aliases_input_registers
from References.Icons.HoldingRegisters import holding_registers as icons_holding_registers
from References.Icons.InputRegisters import input_registers as icons_input_registers
from Data.Vars import get_description, get_by_name, replace_options, get_key_from_value
import configparser
from Data.Logger import logger

configfile = "config.ini"
config = configparser.ConfigParser()
config.read(configfile)
# TCP auto connect on first modbus request
try:
    c = ModbusClient(host=config["Vent"]["ip"], port=502, auto_open=True, auto_close=True, timeout=1)
except Exception as e:
    logger.error("MODBUS: Unable to connect to Vent - {}".format(e))
    exit(1)


class ModBus:
    @staticmethod
    def get_coils():
        output = {}
        coils = c.read_coils(0, 25)
        try:
            coils_len = len(coils)
        except TypeError:
            logger.error("MODBUS: Unable to connect to Vent")
            return output
        for coil in range(coils_len):
            output[get_description(variables_coils, coil)] = {"value": coils[coil],
                                                              "description": get_description(description_coils, coil),
                                                              "unit": None,
                                                              "control": get_description(control_coils, coil)}
        return output

    @staticmethod
    def get_discrete_inputs():
        output = {}
        discrete_inputs = c.read_discrete_inputs(0, 19)
        try:
            discrete_inputs_len = len(discrete_inputs)
        except TypeError:
            logger.error("MODBUS: Unable to connect to Vent")
            return output
        for discrete_input in range(discrete_inputs_len):
            output[get_description(variables_discrete_inputs, discrete_input)] = {
                "value": discrete_inputs[discrete_input],
                "description": get_description(
                    description_discrete_inputs, discrete_input),
                "unit": None}
        return output

    @staticmethod
    def get_alarms():
        output = {}
        alarms = c.read_discrete_inputs(19, 44)
        try:
            alarms_len = len(alarms)
        except TypeError:
            logger.error("MODBUS: Unable to connect to Vent")
            return output
        for alarm in range(alarms_len):
            description = get_description(description_alarms, alarm)
            output[get_description(variables_alarms, alarm)] = {"value": alarms[alarm],
                                                                "description": description,
                                                                "unit": None,
                                                                "icon": "mdi:alarm-light"}
        return output

    @staticmethod
    def get_input_registers():
        output = {}
        input_registers = c.read_input_registers(0, 51)
        try:
            input_registers_len = len(input_registers)
        except TypeError:
            logger.error("MODBUS: Unable to connect to Vent")
            return output
        for input_register in range(input_registers_len):
            unit = get_description(units_input_registers, input_register)
            if unit == "°C":
                if int(input_registers[input_register] / 10) > 6000:
                    value = float((input_registers[input_register] - 65540) / 10)
                else:
                    value = float(input_registers[input_register] / 10)
            elif unit == "h":
                hours = (input_registers[input_register] >> 8) & 0xFF
                minutes = input_registers[input_register] & 0xFF
                days = input_registers[input_register + 1]
                value = int((days * 24) + hours + (minutes / 60))
            else:
                value = input_registers[input_register]
            if value not in (32768, 3276.8):
                if input_register not in (26, 28, 30, 35, 36):
                    aliases = get_description(aliases_input_registers, input_register)
                    icon = get_description(icons_input_registers, input_register)
                    description = get_description(description_input_registers, input_register)
                    if aliases is not None:
                        value = aliases.get(str(value))
                    output[get_description(variables_input_registers, input_register)] = {"value": value,
                                                                                          "description": description,
                                                                                          "unit": unit,
                                                                                          "icon": icon}
        return output

    @staticmethod
    def get_holding_registers():
        output = {}
        holding_registers = c.read_holding_registers(0, 76)
        try:
            holding_registers_len = len(holding_registers)
        except TypeError:
            logger.error("MODBUS: Unable to connect to Vent")
            return output
        for holding_register in range(holding_registers_len):
            unit = get_description(units_holding_registers, holding_register)
            if unit == "°C" and holding_register != 44:
                value = float(holding_registers[holding_register] / 10)
            else:
                value = holding_registers[holding_register]
            if value not in (32768, 3276.8):
                if holding_register not in range(27, 37):
                    controls = get_description(control_holding_registers, holding_register)
                    aliases = get_description(aliases_holding_registers, holding_register)
                    description = get_description(description_holding_registers, holding_register)
                    icon = get_description(icons_holding_registers, holding_register)
                    if aliases is not None:
                        value = aliases.get(str(value))
                    if aliases is not None and controls is not None:
                        control = replace_options(controls, aliases)
                    else:
                        control = controls
                    output[get_description(variables_holding_registers, holding_register)] = {"value": value,
                                                                                              "description": description,
                                                                                              "unit": unit,
                                                                                              "control": control,
                                                                                              "icon": icon}
        return output

    @staticmethod
    def post_holding_registers(key, value):
        address = get_by_name(variables_holding_registers, key)
        aliases = get_description(aliases_holding_registers, address)
        if aliases is not None:
            value = get_key_from_value(aliases, value)
        c.write_single_register(address, int(value))
        return True

    @staticmethod
    def post_coils(key, value):
        address = get_by_name(variables_coils, key)
        c.write_single_coil(address, bool(value))
        return True
